### ISSUE TITLE
[HOLD for payment 2024-07-17] [$125] Invoice-User B's workspace appears in User A's workspace list after paying invoice as business

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#44811]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.4-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
Precondition:
- User A and User B have their own workspaces.

1. Go to staging.new.expensify.com
2. [User A] Send invoice to User B.
3. [User B] Go to invoice chat.
4. [User B] Click on the pay button > Pay as a business > Pay elsewhere.
5. [User A] Go to Account settings > Workspaces.

## Expected Result:
User B's workspace will not appear in User A's workspace list.

## Actual Result:
User B's workspace appears in User A's workspace list after paying invoice as a business.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/6f68a2ec-b1b9-47a5-bcb7-b5411cc086d5

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011ae1abcb3755a5ef</li>
        <li>Upwork Job ID: 1808845128155915041</li>
        <li>Last Price Increase: 2024-07-04</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @kevinksullivan</details>

**Comment by izarutskaya** [comment]
@kevinksullivan FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by izarutskaya** [comment]
We think this issue might be related to the #vip-vsb.

**Comment by izarutskaya** [comment]
Production

![image (2)](https://github.com/Expensify/App/assets/115492554/fc6867f2-1116-426f-a7e9-89a6b1a7b819)

**Comment by flodnv** [comment]
@cristipaval I think you've been working on this, can you take a look please?

**Comment by cristipaval** [comment]
The App needs to filter out the workspaces in which the user has no role. By the time when we [decided](https://expensify.slack.com/archives/CSL3XBCCR/p1712135562728899) to save public info about receiver policies in the Onyx for all participants, the App was filtering those workspaces out (see [this](https://expensify.slack.com/archives/CSL3XBCCR/p1712766630501319?thread_ts=1712135562.728899&cid=CSL3XBCCR)).

**Comment by cristipaval** [comment]
The solution looks good to me but I think that the bounty for this issue should be 125 USD, given that I also posted the solution in plain English [here](https://github.com/Expensify/App/issues/44811#issuecomment-2208633695)

**Comment by cristipaval** [comment]
@nyomanjyotisa please go ahead and open a PR asap 🙏

**Comment by mountiny** [comment]
Pending a fix validation in staging

**Comment by cristipaval** [comment]
> I believe the payment was already due as of 2024-07-17

You're right. @kevinksullivan could you please handle the payments? 🙏

**Comment by kevinksullivan** [comment]
Offer sent to @nyomanjyotisa . @eVoloshchak can you complete the checklist please?

**Comment by kevinksullivan** [comment]
Payment summary:
- @eVoloshchak $150 (ND)
- @nyomanjyotisa $150

**Comment by kevinksullivan** [comment]
Paid. Waiting on checklist @eVoloshchak

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/33280
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/33280/files#r1686570145
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: I don't think an additional discussion is needed as this is a minor edge case

**Regression Test Proposal**

Prerequisite: User A and User B have their own workspaces.

1. [User A] Send invoice to User B.
2. [User B] Go to invoice chat.
3. [User B] Click on the pay button > Pay as a business > Pay elsewhere.
4. [User A] Go to Account settings > Workspaces.
5. Verify that User B's workspace does not appear in the workspace list

**Do we agree 👍 or 👎**

**Comment by JmillsExpensify** [comment]
$150 approved for @eVoloshchak

**Comment by kevinksullivan** [comment]
adding regression test and closing out!
